checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -223,6 +253,10 @@ protected void refineGT(
             knownToBeLessThanLengthOf(array, right, store, in);
         }
 
+        if (right instanceof NumericalAdditionNode || right instanceof NumericalSubtractionNode) {
+            propagateToSubterms((BinaryOperationNode) right, refinedRight, store);
```

Comment:
I looked through the code with an eye for reducing duplicated code, and I think that the code as-is is fine. In particular, all the methods that `knownToBeLessThanLengthOf` calls rely on the property that it is, in fact, known to be less than length of. For instance, the version of `knownToBeArrayLength` that takes a `NumericalAdditionNode` is able to account for lowerbound annotations because of that property.